torch/nn/utils/parametrize.py

```diff
@@ -0,0 +1,375 @@
+        r""" It modifies the current parametrization, chaining a new one to it.
+        We have that ``p2.evaluate(p1.evaluate(t)) == p2.chain(p1).evaluate(t)``
+
+        The original parametrization is modified after chain is applied to it
+
+        Args:
+            parametrization (nn.Parametrization): the parametrization to be chained
+        Returns:
+            Module: The original module with the parametrization module appended to it
+        """
+        if not isinstance(parametrization, Parametrization):
+            raise ValueError("The object '{}' is not a Parametrization"
+                             .format(parametrization))
+
+        if self.is_registered() and parametrization.is_registered():
+            raise ValueError("Both parametrizations are registered to a tensor. "
+                             "The chaining is ambiguous.")
+        last = self.last_parametrization()
+        is_parametrized = hasattr(last, "orig")
+        if is_parametrized:
+            # We move the orig tensor to the new last parametrization
+            orig = last.orig
+            delattr(last, "orig")
+
+        _set_parametrization(last, parametrization, "orig", set_cache=False)
+        if is_parametrized:
+            new_last = parametrization.last_parametrization()
+            if isinstance(orig, Parameter):
+                new_last.register_parameter("orig", orig)
+            else:
+                new_last.register_buffer("orig", orig)
+        return self
+
+    def is_chained(self):
+        # TODO Write docs
+        return is_parametrized(self, "orig")
+
+    def is_registered(self):
+        # TODO Write docs
+        try:
+            self.original()
+            return True
+        except ValueError:
+            return False
+
+
+def _set_parametrization(module, parametrization, tensor_name, set_cache):
+    r""" Sets up the parametrization mechanism used by parametrizations.
+    This works by substituting the class of the module by a class
+    that extends it and makes `tensor_name` into a property. It also
+    regsters the parametrization under `tensor_name + "_param"`.
+    If ``set_cache == True``, it also sets up the caching mechanism
+    """
+
+    # Define the getter and set the caching mechanism
+    if set_cache:
+        def get_parametrized(module):
+            cache = getattr(module, tensor_name + "_cache")
+            if cache is None:
+                param = getattr(module, tensor_name + "_param")
+                return param.evaluate()
+            else:
+                return cache
+    else:
+        # Simplified version of the above
+        def get_parametrized(module):
+            param = getattr(module, tensor_name + "_param")
+            return param.evaluate()
+
+    if not is_parametrized(module):
+        # If it has not been parametrized, we create a new class so that
+        # we can inject properties in it
+        param_cls = type("Parametrized" + module.__class__.__name__,
```

Comment:
Something to check before this gets merged: can parametrized models still get `torch.save`d? Pickle doesn't like weird things like this too much, so we should verify this.

Reply:
I'm pretty sure it won't work out of the box, but it should be possible to fix it using some custom reducer functions.

Reply:
I have never dealt with these corner cases of the pickling module, so I don't know how to go in this regard. Happy to take suggestion.

Reply:
Can you please try pickling modules that have parametrizations applied and see if that fails or not? We'll think about solutions once we confirm that it is an issue.

Reply:
Fails with:

```
_pickle.PicklingError: Can't pickle <class 'torch.nn.utils.parametrize.ParametrizedLinear'>: attribute lookup ParametrizedLinear on torch.nn.utils.parametrize failed
```

Reply:
In the last commit I gave a go at this. Now parametrized layers work well with `torch.load` / `torch.save`. Take a look at it, see if you like it.

It is somewhat hacky, but one literally needs to hack around the limitations of `pickle`. If you know of a better way I can change it.